man/thin.1.ronn
thin(1) -- high performance Ruby server
=======================================

## SYNOPSIS

`thin` [options] [<config.ru>]<br>
`thin` `-c` <thin.conf.rb> [<config.ru>]

## DESCRIPTION

Thin is a high performance Ruby server. Thin is generally used as an app server for serving [Rack applications](http://rack.rubyforge.org/doc/SPEC.html) behind full blown HTTP servers such as [nginx](http://nginx.org).

## FILES

The `thin` command will load the Rack application defined in a rackup config file. The default behaviour is to look for a file named `config.ru` that lives at the root of the application if the argument is omitted. This file is evaluated in the context of a [Rack::Builder](http://rack.rubyforge.org/doc/Rack/Builder.html) class.

The `thin` command can also take a server configuration file with the `-c` option. See thin-conf(5) for the format of this file.

## OPTIONS

These options control the Ruby environment and are the same as the `ruby` and `rackup` commands.

  * `-e`, `--eval` <LINE>:
    Evaluate a LINE of Ruby code.

  * `-d`, `--debug`:
    Set debugging flags (set $DEBUG to true). This is also output the request and response to stdout.

  * `-w`, `--warn`:
    Turn warnings on for your script.

  * `-I`, `--include` <PATH>:
    Add a path to the $LOAD_PATH. May be used more than once.

  * `-r`, `--require` <LIBRARY>:
    Require the library, before executing your script. May be used more than once.

These options control the Thin server.

  * `-o`, `--host` <HOST>:
    Bind to HOST. Leave blank to bind to any address. Use *[...]* to specify IPv6 addresses.

  * `-p`, `--PORT` <PORT>:
    Bind to PORT.

  * `-E`, `--env` <ENVIRONMENT>:
    Use the Rack ENVIRONMENT.
    
    *deployment*:
      will load the *ContentLength*, *Chunked* and *CommonLogger* middlewares.<br>
    *development*:
      will load the sames middlewares as *deployment* with the addition of *ShowExceptions* and *Lint*.
    
    Any other environment will not load any middleware. This option will also set the *RACK_ENV* environment variable.

  * `-D`, `--daemonize`:
    Run daemonized in the background.

  * `-P`, `--pid` <FILE>:
    File to store the process ID in.

  * `-l`, `--log` <FILE>:
    File to redirect stdout and stderr to.

  * `-c`, `--config` <FILE>:
    Configure the Thin server by evaling this configuration file. See thin-conf(5) for the format of that file.

## SIGNALS

Thin master process can handle the following signals.

  * `WINCH`:
    Gracefully kill all workers but keep master alive

  * `TTIN`:
    Increase number of workers

  * `TTOU`:
    Decrease number of workers

  * `QUIT`:
    Kill workers and master in a graceful way

  * `TERM`, `INT`:
    Kill workers and master immediately

## MODES

By default Thin is a preforking server like Nginx and Unicorn. On systems where fork(2) is not supported, Thin will fallback to running in a single process. While running in this mode, a number of features are not supported.

Here is the list of features that do **NOT** work when running in single process mode.

  * Signals, except `TERM` and `INT`.
  * Run daemonized in the background (`-D` option).
  * Log file (`-l` option).

Single process mode can also be forced by setting the number of worker processes to *0*.

## COPYRIGHT

Thin is Copyright (C) 2012 Marc-Andre Cournoyer <http://macournoyer.com/>

## SEE ALSO

thin-conf(5), [Rack spec](http://rack.rubyforge.org/doc/SPEC.html)

CHANGELOG
== 1.4.1 Chromeo Fix
 * Fix error when sending USR1 signal and no log file is supplied.

== 1.4.0 Chromeo
 * kill -USR1 $PID for log rotation [catwell].
 * Fix HUP signal being reseted after deamonization [atotic].
 * Fix error with nil addresses in Connection#socket_address.

== 1.3.2 Low-bar Squat
 * Remove mack and halcyon Rack adapters from automatic detection.

== 1.3.1 Triple Espresso
 * Fix service not working pre 1.9.

== 1.3.0 Double Espresso
 * BREAKING CHANGE: Thin no longer ships with fat Windows binaries.
   From now on, to install on Windows, install https://github.com/oneclick/rubyinstaller/wiki/Development-Kit.
 * BREAKING CHANGE: Remove automatic Content-Length setting.
   It is now the responsibility of the app (or a middleware) to set the Content-Length.
 * Log errors to STDERR [textgoeshere]
 * Shut down gracefully when receiving SIGTERM [ddollar]
     
     Processes are allowed a chance to shut down gracefully when receiving
     SIGTERM (http://en.wikipedia.org/wiki/SIGTERM).
     
     On Heroku, when shutting down a process, we send a SIGTERM followed 10
     seconds later with a SIGKILL, similar to the behavior of the init daemon
     on most Unix systems. This patch will allow Heroku apps to shut down
     gracefully when they need to be terminated / moved.

== 1.2.11 Bat-Shit Crazy
 * Fix pure Ruby gem to not include binary.

== 1.2.10 I'm dumb (BAD RELEASE, DON'T USE)
 * I really am (bad release fix)

== 1.2.9 Black Keys Extra Plus Wow (BAD RELEASE, DON'T USE)
 * Improve fat binary loading.

== 1.2.8 Black Keys
 * Allow the connection to remain open for 1xx statuses [timshadel]

     Both the 100 and 101 status codes require that the connection to the
     server remain open. The 100 status code is used to tell the client that
     the server is still receiving its request, and will continue to read
     request input on the connection. The 101 status code is used to upgrade
     the existing connection to another protocol, and specifically is NOT
     used to upgrade a separate connection. Therefore, the connection must
     remain open after this response in order to facilitate that.
   
 * Accept IE7 badly encoded URL (eg.: %uEEEE)
 * Fix gemspec to work w/ Bundler [smparkes]
 * Add SSL support [tmm1]
 * Catch Errno::EPERM in Process.running? [Tony Kemp]
   On some systems (e.g. OpenBSD) you receive an EPERM exception if
   you try to Process.getpgid on a process you do not own (even if you
   are root). But it does mean that the process exists, so return true.
 * Fix Rails version check that select which Rack adapter to use. Was using CGI adapter in Rails 3.
 * Ignore SIGHUP when no restart block is given
 * Add SSL options to thin command line tool [goldmann]

      --ssl                    Enables SSL
      --ssl-key-file PATH      Path to private key
      --ssl-cert-file PATH     Path to certificate
      --ssl-verify             Enables SSL certificate verification
    
  * Expose peer SSL certificate in env (rack.peer_cert) [fd]
  * Adjusting unix socket permissions to be more open [mbj]
  
== 1.2.7 No Hup
 * Support multiple Ruby version (fat binaries under windows)
 * Do not trap unsupported HUP signal on Windows

== 1.2.6 Crazy Delicious
 * Make work with Rails 3 out-of-the-box.
 * Auto-detect and load config.ru files on start. Makes Rails 3 work.
 * Fix signals being ignored under 1.9 when daemonized.

== 1.2.5 This Is Not A Web Server
 * Add rolling restart support (--onebyone option) [sikachu]
 * Force external_encoding of request's body to ASCII_8BIT [jeremyz]
 * Ensure Rack base API is used in Rails adapter only if version >= 2.3.2 [#111 state:resolved]

== 1.2.4 Flaming Astroboy
 * Fix a few issues in thin to make it a better "gem citizen" [josh]
 * Fix test for rack based Rails in adapter under Ruby >= 1.8.7 [#109 state:resolved]
 * Fix Remote address spoofing vulnerability in Connection#remote_address [Alexey Borzenkov]
 * Fix uninitialized constant ActionController::Dispatcher error with Rails 1.2.3 [Chris Anderton] [#103 state:resolved]

== 1.2.2 I Find Your Lack of Sauce Disturbing release
 * Fix force kill under 1.9 [Alexey Chebotar]
 * Fix regression when --only option is used w/ --socket.
 * Add process name 'tag' functionality. Easier to distinguish thin daemons 
   from eachother in process listing [ctcherry]

== 1.2.1 Asynctilicious Ultra Supreme release
 * Require Rack 1.0.0
 * Require EventMachine 0.12.6
 * Use Rails Rack based dispatcher when available
 * Allow String for response body
 * Require openssl before eventmachine to prevent crash in 1.9

== 1.2.0 Asynctilicious Supreme release 
 * Add support for Windows mingw Ruby distro [Juan C. Rodriguez]
 * Add async response support, see example/async_*.ru [raggi]

== 1.1.1 Super Disco Power Plus release 
 * Fix bug when running with only options [hasimo]

== 1.1.0 Super Disco Power release
 * Require EventMachine 0.12.4
 * Remove Thin handler, now part of Rack 0.9.1
 * Fix Rack protocol version to 0.1 in environment hash.
 * Fix error when passing no_epoll option to a cluster.
 * Omit parsing #defined strings [Jérémy Zurcher]
 * Defaults SERVER_NAME to localhost like webrick does [#87 state:resolved]
 * Namespace parser to prevent error when mongrel is required [cliffmoon]
 * Set RACK_ENV based on environment option when loading rackup file [Curtis Summers] [#83 state:resolved]
 * Fixes a warning RE relative_url_root when using a prefix with Rails 2.1.1 [seriph] [#85 state:resolved]
 * --only can work as a sequence number (if < 80) or a port number (if >= 80) [jmay] [#81 state:resolved]

== 1.0.0 That's What She Said release
 * Fixed vlad.rake to allow TCP or socket [hellekin]
 * Updated Mack adapter to handle both <0.8.0 and >0.8.0 [Mark Bates]
 * rails rack adapter uses File.readable_real? so it recognizes ACL permissions [Ricardo Chimal]
 * Log a warning if Rack application returns nil body [Michael S. Klishin]

site/thin.rb
require 'rubygems'
require 'atchoum'

class Thin < Atchoum::Website
  ROOT = ENV['SITE_ROOT'] || ''
  
  def layout
    xhtml_html do
      head do
        title 'Thin - yet another web server'
        link :rel => 'stylesheet', :href => "#{ROOT}/style.css", :type => 'text/css'
      end
      body do
        ul.menu! do
          li { a 'about', :href => "#{ROOT}/" }
          li { a 'download', :href => "#{ROOT}/download/" }
          li { a 'usage', :href => "#{ROOT}/usage/" }
          li { a 'doc', :href => "#{ROOT}/doc/" }
          li { a 'code', :href => "http://github.com/macournoyer/thin/" }
          li { a 'bugs', :href => "http://github.com/macournoyer/thin/issues" }
          li { a 'users', :href => "#{ROOT}/users/" }
          li { a 'community', :href => "http://groups.google.com/group/thin-ruby/" }
        end

        div.container! do
          div.header! do
            a(:href => :index, :title => 'home') do
              img.logo! :src => "#{ROOT}/images/logo.gif", :alt => 'Thin'
            end
            h2.tag_line! "A fast and very simple Ruby web server"
          end
        
          div.content! do
            self << yield
          end

          div.footer! do
            hr
            text "&copy; "
            a(:href => 'http://macournoyer.com') { 'Marc-Andr&eacute; Cournoyer' }
          end
        end
      end
    end
  end
  
  def index_page
    h2 'What'
    p 'Thin is a Ruby web server that glues together 3 of the best Ruby libraries in web history:'
    ul do
      li do
        text "the #{a 'Mongrel parser', :href => 'http://www.zedshaw.com/tips/ragel_state_charts.html'}"
        text ", the root of #{a 'Mongrel', :href => 'http://mongrel.rubyforge.org/'} speed and security"
      end
      li do
        a "Event Machine", :href => 'http://rubyeventmachine.com/'
        text ", a network I/O library with extremely high scalability, performance and stability"
      end
      li do
        a "Rack", :href => 'http://rack.rubyforge.org/'
        text ", a minimal interface between webservers and Ruby frameworks"
      end
    end
    p <<-EOS
      Which makes it, with all humility, the most secure, stable, fast and extensible Ruby web server
      bundled in an easy to use gem for your own pleasure.
    EOS
    
    h2 'Why'
    div.graph do
      h3 'Request / seconds comparison'
      img :alt => 'Chart', :title => 'Benchmarks', :src => chart_url([
        ['WEBrick',     [297.37, 296.65, 297.16]],
        ['Mongrel',     [556.67, 622.90, 428.23]],
        ['Evented M.',  [517.97, 657.89, 656.17]],
        ['Thin',        [719.53, 782.11, 776.40]]
      ], :min => 200, :scale => 6.5, :size => '350x150', :width => 16,
         :colors => %w(000000 666666 cccccc),
         :legends => ['1 c req.', '10 c req.', '100 c req.'])
      em 'c req. = Concurrency Level (number of simultaneous requests)'
    end
    
    h2 'How'
    pre 'sudo gem install thin'

    p 'Go to your Rails app directory and run:'
    pre 'thin start'
  end
  
  def download_page
    h2 'Install the Gem'

    p 'To install the latest stable gem'
    pre 'sudo gem install thin'
    
    h2 'Install from source'

    p { "Clone the #{a 'Git repository', :href => 'http://github.com/macournoyer/thin/'}" }
    pre "git clone git://github.com/macournoyer/thin.git"
    
    p 'Hack the code, patch it, whatever, build the Gem and install'
    pre "cd thin\nsudo rake install"
  end
  
  def usage_page
    h2 'Using with Rails'

    p { "After installing the Gem, a #{code 'thin'} script should be in your " +
        "path to easily start your Rails application." }

    pre "cd to/your/rails/app\nthin start"
    
    h2 'Using with anything, ANYTHING!'
    
    p "But Thin can also load Rack config file so you can use it with any framework that " +
      "supports Rack. Even your own that is, like, soooo much better then Rails, rly!"

    em.filename 'fart.ru'
    pre <<-EOS.gsub(/^\s{6}/, '')
      app = proc do |env|
        [
          200,          # Status code
          {             # Response headers
            'Content-Type' => 'text/html',
            'Content-Length' => '2',
          },
          ['hi']        # Response body
        ]
      end
      
      # You can install Rack middlewares
      # to do some crazy stuff like logging,
      # filtering, auth or build your own.
      use Rack::CommonLogger
      
      run app
    EOS
    
    pre 'thin start -R fart.ru'
    p { "See #{a 'Rack doc', :href => 'http://rack.rubyforge.org/doc/'} for more." }
    
    hr
    
    h2 'Deploying'
    
    p 'Deploying a cluster of Thins is super easy. Just specify the number of servers you ' + 
      'want to launch.'
    pre 'thin start --servers 3'
    
    p 'You can also install Thin as a runlevel script (under /etc/init.d/thin) that will start all your servers after boot.'
    pre 'sudo thin install'
    
    p 'and setup a config file for each app you want to start:'
    pre 'thin config -C /etc/thin/myapp.yml -c /var/...'
    
    p { "Run #{code 'thin -h'} to get all options." }
    
    h2 'Behind Nginx'
    
    p { "Check out this #{a 'sample Nginx config', :href => 'http://brainspl.at/nginx.conf.txt'} " +
        "file to proxy requests to a Thin backend." }
    
    p 'then start your Thin cluster like this:'
    pre 'thin start -s3 -p 5000'
    
    p 'You can also setup a Thin config file and use it to control your cluster:'
    pre "thin config -C myapp.yml -s3 -p 5000\nthin start -C myapp.yml"
    
    p 'To connect to Nginx using UNIX domain sockets edit the upstream block in your nginx config file:'
    
    em.filename 'nginx.conf'
    pre <<-EOS.gsub(/^\s{6}/, '')
      upstream  backend {
         server   unix:/tmp/thin.0.sock;
         server   unix:/tmp/thin.1.sock;
         server   unix:/tmp/thin.2.sock;
      }
    EOS
    
    p 'and start your cluster like this:'
    pre 'thin start -s3 --socket /tmp/thin.sock'
  end
  
  def users_page
    h2 'Users'
    
    p "Who's using Thin ?"
    
    ul do
      li { a 'Heroku', :href => 'http://heroku.com/' }
      li { a 'HostingRails', :href => 'http://www.hostingrails.com/mongrel_and_thin_hosting' }
      li { a "Audiofarm.org", :href => 'http://audiofarm.org'}
      li { a "What The Movie", :href => 'http://whatthemovie.com/'}
      li { a 'Mobile Defense', :href => 'https://www.mobiledefense.com' }
      li { a 'RefactorMyCode.com', :href => 'http://refactormycode.com/' }
      li { a "Standout Jobs", :href => 'http://standoutjobs.com/'}
      li { a "Gitorious", :href => 'http://gitorious.org/' }
      li { a "Kevin Williams Blog", :href => 'http://www.almostserio.us/articles/tag/thin' }
      li { a "Dinooz", :href => 'http://www.nicomoayudarte.com/' }
      li { a "Mobile Dyne Systems", :href => 'http://www.mobiledyne.com/' }
      li { a "feelfree.homelinux.com", :href => 'http://feelfree.homelinux.com' }
      li { a "to2posts.com", :href => 'http://to2blogs.com/' }
      li { a "James on Software", :href => 'http://jamesgolick.com/' }
      li { a "FlashDen", :href => 'http://flashden.net/' }
      li { a "Calico Web Development", :href => 'http://www.calicowebdev.com/' }
      li { a "Cornerstone Web", :href => 'http://www.cornerstoneweb.org/' }
      li { a "Osmos", :href => 'http://www.getosmos.com/' }
      li { a "Lipomics", :href => 'http://www.lipomics.com/' }
      li { a "Joao Pedrosa's Blog", :href => 'http://www.deze9.com/jp/blog/post?p=enabling-thin-support-for-this-site-replacing' }
      li { a "RaPlanet", :href => 'http://planet.zhekov.net/' }
      li { a "Ninja Hideout blog", :href => 'http://blog.ninjahideout.com/' }
      li { a "Hoodows Blog", :href => 'http://blog.hoodow.de/articles/2008/02/09/thin' }
      li { a "moonitor.org", :href => 'http://moonitor.org' }
      li { a "Daniel Fischer's Blog", :href => 'http://www.danielfischer.com/' }
      li { a "RoundHaus", :href => 'http://www.roundhaus.com/' }
      li { a "Socks and Sandals", :href => 'http://blog.cbcg.net/' }
      li { a "indiagoes", :href => 'http://www.indiagoes.com/' }
      li { a "ajaxwhois", :href => 'http://ajaxwhois.com/' }
      li { a "Sproglogs", :href => 'http://sproglogs.com/' }
      li { a "Look to the Stars", :href => 'http://www.looktothestars.org/' }
      li { a "kluster", :href => 'http://kluster.com'}
      li { a "SocialSpark", :href => 'http://socialspark.com'}
      li { a "Tanga.com", :href => 'http://tanga.com'}
      li { a "Opening Times", :href => 'http://opening-times.co.uk/'}
      li { a "p0pulist", :href => 'http://p0pulist.com/'}
      li { a "pickhost.eu", :href => 'http://pickhost.eu'}
      li { a "HowFlow", :href => 'http://howflow.com/'}
      li { a "boo-widgets", :href => 'http://boo-box.com/site/en/widget/get'}
      li { a "BigCity.cz", :href => 'http://www.bigcity.cz/'}
      li { a "yabadaba.ru", :href => 'http://yabadaba.ru/'}
      li { a "Freebootr.com", :href => 'http://freebootr.com/'}
      li { a "Hacknight", :href => 'http://hacknight.gangplankhq.com/'}
      li { a "Sizzix", :href => 'http://www.sizzix.com/'}
      li { a "Ellison Retailers", :href => 'http://www.ellisonretailers.com/'}
      li { a "ShareMeme", :href => 'http://sharememe.com/'}
      li { a "ylastic", :href => 'http://ylastic.com/'}
      li { a "hello2morrow", :href => 'http://www.hello2morrow.com/'}
      li { a "TimmyOnTime", :href => 'http://www.timmyontime.com/'}
      li { a "Generous", :href => 'http://generous.org.uk/'}
      li { a "Catapult Magazine", :href => 'http://www.catapultmagazine.com/'}
      li { a "Good-Tutorials", :href => 'http://www.good-tutorials.com/'}
      li { a "audioreload", :href => 'http://www.audioreload.com/'}
    end
    
    p { "If you'd like to have your site listed here, #{a 'drop me an email', :href => 'mailto:macournoyer@gmail.com'}" }
  end
  
  private
  
    def chart_url(data, options={})
      size   = options[:size]  || '250x150'
      width  = options[:width] || 50
      min    = options[:min]   || 0
      scale  = options[:scale] || 1.0
      colors = options[:colors] || '000000'
      legends = options[:legends]
    
      data_matrix  = data.collect { |d| Array(d[1]) }
      chart_matrix = data_matrix[0].zip(*data_matrix[1..-1]) # transpose the matrix
      chart_data   = chart_matrix.collect { |c| c.collect { |d| (d - min) / scale }.join(',') }.join('|')
      labels       = data.collect{|d|d[0]}.join('|')
    
      "http://chart.apis.google.com/chart?cht=bvg&chd=t:#{chart_data}&chbh=#{width}&chs=#{size}&chl=#{labels}&chco=#{colors.join(',')}&chdl=#{legends.join('|')}"
    end
end